keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,380 @@
+the PVC. This can be mitigated by initially introducing the fix behind a feature gate, followed by explicitly deprecating
+the current behavior.
+
+## Design Details
+
+A feature gate named `HonorPVReclaimPolicy` will be introduced for both `kube-controller-manager` and `external-provisioner`.
+
+A new finailzer `PVDeletionProtectionFinalizer` will be added in syncVolume() of sig-storage-lib-external-provisioner/controller/controller.go if needed. The controller will find the corresponding StorageClass from the PV and find the provisioner specified in the StorageClass. The new finalizer will be added only if the provisioner in the StorageClass matches the provisioner name of the external-provisioner. This means statically provisioned PVs won't get this finalizer unless if the correct storageClassName is provided in the PV. Also PVs provisioned by in-tree volume plugins won't get this new finalizer. We can add this new finalizer for in-tree volume plugins later if needed in the future.
+
+```go
+// PVDeletionProtectionFinalizer finalizer is added to the PV when the controller sees it
+PVDeletionProtectionFinalizer = "kubernetes.io/pv-deletion-protection"
+```
+
+In `deleteVolumeOperation` in kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go, PV without `PVDeletionProtectionFinalizer` will be skipped as it is already processed and PV with `PVDeletionProtectionFinalizer` will proceed with volume deletion. This will allow external-provisioner to handle the deletion of the CSI volumes. The finalizer `PVDeletionProtectionFinalizer` will only be deleted after the underlying CSI volume is deleted.
+
+```go
+// deleteVolumeOperation deletes a volume. This method is running in standalone
+// goroutine and already has all necessary locks.
+func (ctrl *PersistentVolumeController) deleteVolumeOperation(volume *v1.PersistentVolume) (string, error) {
+	klog.V(4).Infof("deleteVolumeOperation [%s] started", volume.Name)
+	newVolume, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Get(context.TODO(), volume.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(3).Infof("error reading persistent volume %q: %v", volume.Name, err)
+		return "", nil
+	}
+
+	// PV without PVDeletionProtectionFinalizer was already processed and its volume deleted.
+	if !ctrl.hasPVDeletionProtectionFinalizer(volume, pvutil.PVDeletionProtectionFinalizer) {
+		if newVolume.GetDeletionTimestamp() != nil {
+			klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+			return "", nil
+		}
+	}
+	//.....
+	//.....
+	//.....
+	klog.V(4).Infof("deleteVolumeOperation [%s]: success", volume.Name)
+        return pluginName, err
+    }
+// Remaining code below skipped for brevity...
+// Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+
+On the driver side, additional checks are made whenever `DeletionTimestamp` is set.
```

Comment:
I think no changes should be needed on the external-provisioner or driver side. It already handles external-provisioner.volume.kubernetes.io/finalizer.

Reply:
Correction: the library adds the finalizer only to newly provisioned PVs. We should extend the library to (optionally) add the finalizer to all PVs (that are handled by the external-provisioner), to have all PVs protected once the feature is enabled.

Reply:
Added.